File: go/bob/bob.go
package bob

import "strings"

const testVersion = 3

func Hey(s string) string {

	s = strings.Trim(s, " ")
	alphabet := "abcdefghijklmnopqrstuvwxyz"
	switch {
	case strings.ToUpper(s) == s && strings.ContainsAny(strings.ToLower(s), alphabet):
		return "Whoa, chill out!"
	case len(s) > 0 && strings.EqualFold(string(s[len(s)-1:]), "?"):
		return "Sure."
	case !strings.ContainsAny(s, "abcdefghijklmnopqrstuvwxyz1234567890"):
		return "Fine. Be that way!"
	default:
		return "Whatever."
	}
}

File: haskell/bob/src/Bob.hs
module Bob (responseFor) where

import Data.Char

responseFor :: String -> String
responseFor xs 
    | isYelling xs && '?' == last xs  = "Calm down, I know what I'm doing!"
    | isYelling xs = "Whoa, chill out!"
    | not ( all isSpace xs) && '?' == last (filter (not . isSpace) xs) = "Sure."
    | not $ any isAlphaNum xs  = "Fine. Be that way!"
    | otherwise = "Whatever."
    where isYelling xs = all (\c -> c `notElem` ['a'..'z'])  xs && any (\c -> c `elem` ['A'..'Z']) xs
            
